### DISCUSSION TITLE
Advice for Nozzle Priming

### DISCUSSION THREAD

**Discussion by ferm10n** [smartin015/continuousprint#202] [Q&A]
In the Printer Settings of my prusaslicer profile, the Start G-code has commands to prime the nozzle: heat up the nozzle, then extrude a thick line outside the print area. (this is a default behavior)

This helps ensure molten filament is at the tip of the nozzle when the print starts, but when it's used with CPQ printing many parts, this prime line can get stuck on the nozzle and ruin a print's first layer.

I'm just wondering if any users of this plugin have found a way to deal with this?

One idea I had was to use [my scripting approach](https://github.com/smartin015/continuousprint/discussions/136) to make the location of the prime line be in a new location each print.

**Comment by smartin015** [comment]
Hey @ferm10n - I'd love to see a workable answer to this as well, albeit as a reference for other cartesian printer users. 

Have you considered having your purge line "tag" the print you're making, so it lifts up along with the print? A potential way to quickly try this would be to slice your part with a brim - assuming the brim prints first and is connected, it should act as a purge line of sorts and you can remove the default starting purge line behavior.

**Comment by rogiepoo** [comment]
Hi. I put my primeline to the front right. My bed clearing script works from right to left. I tend to get duplicate, sometimes triplicate primelines piled up, but for some reason they don't get stuck on my prints. I like the idea of integrating it with a part. 
r

**Comment by flapwazzle** [comment]
I'm about to start a print farm and print one box until the spool is empty and just collect the boxes every 2 days or so.
Already figured that this would be a problem, the primeline integrating into the object is a workaround but than requires little bit more work afterwards for me and is technically not elegant.
Is it possible to shift the origin of the coordinates of the printer in software? If so we could just shift the origin of the printer by 0.5mm up the Y-Axis and then run easy peasy prints until the spool is empty, reset the printer, clean all the prime lines and start anew.

**Reply by smartin015** [reply]
CPQ's preprocessors would work for shifting origin coordinates, but there's no persisted data between events. That's mostly for simplicity, as IMO allowing a run of a script to affect future script runs makes it way harder to reason about. 

*However* - it looks like you could do something fancy if your printer uses Marlin: [G60](https://marlinfw.org/docs/gcode/G060.html) and [G61](https://marlinfw.org/docs/gcode/G061.html) allow storing and returning to a position, so you could probably do something like this (WARNING UNTESTED):

```
G61 ; Return to purge position
G91 ; Set to relative positioning
G21 ; Set millimeter units
G0 Y0.5 ; Move 0.5mm relative in Y axis
G60 ; Save purge position
<purge line code goes here, probably G0 X# E# etc. etc.>
```

As an added bonus, this just affects the purge line and doesn't affect the print location at all. Note that you'd need to run a script to reset the position back to where it was previously:

```
G21 ; mm units
G90 ; absolute positioning
G0 X# Y# Z0;
```

Otherwise you'll just keep shifting over indefinitely. Also looks like offsets are set in SRAM, so you'd need to run the reset script any time the printer restarts. This may run afoul of CPQ's auto reconnection setting (e.g. reconnecting to the printer and continuing the queue if the printer resets) so be careful if you have that option enabled. 

I'd be interested to hear if this works for you or if you find another solution. "Solving" purge lines would be fantastic.

**Reply by flapwazzle** [reply]
Thanks for you fast and elaborate reply! I run a prusa mk3 and will get 3 prusa mini+, they do run on marlin so I guess that would work.
honestly it wouldnt be too bad if the position of the box would shift with the purge line, because then the satin sheet gets the stress on different spots over time instead of the same one over and over.

Do I understand you correct that this code would be implemented in the start g-code of the machine and should work even tho the printer auto homes before every print? Or should I remove the auto home function and just home once when I use a new spool and reboot the printer when the spool is empty?

**Reply by smartin015** [reply]
Since the purge line code is usually added by the slicer to the gcode file, I'd suggest:

1. Copy the existing purge line code to a "print start" script in CPQ, and delete it from your slicer config
2. Modify the "print start" script, adding the G60/G61 code in front of it to set the purge line position
3. Test thoroughly and carefully ;)

Unfortunately this does mean the purge line won't happen if CPQ is not involved, but it sounds like you'll have a pretty static setup so that shouldn't be too much of a problem. 

Also, I forgot a `G60` at the end of the "reset" script above - that would save the position for later recall in the print start script. And if you have problems with the print position getting overwritten somehow,  the `G60` link above details how to use different save slots so you can save multiple positions or avoid getting overwritten.

HTH

**Reply by smartin015** [reply]
Realizing I didn't answer your other question - autohoming shouldn't affect the saved position, although again I've not actually used this command so YMMV. You could remove it if you want, but I'd suggest trying with it still in place first to see if it's even necessary.

**Reply by flapwazzle** [reply]
So I've looked into the GCode and to me (completely unexperienced with GCode) it looks like the printer goes to a static position: (Y=-3)

` ; Prime line routine

M300 S40 P10 ; chirp
M117 Printing prime line
G0 Z0.15 ; Restore nozzle position - (thanks tim.m30)
M900 K0; Disable Linear Advance for prime line
G92 E0.0 ; reset extrusion distance
G1 Y-3.0 F1000.0 ; go outside print area
G1 E2 F1000 ; de-retract and push ooze
G1 X20.0 E6  F1000.0 ; fat 20mm intro line @ 0.30
G1 X60.0 E3.2  F1000.0 ; thin +40mm intro line @ 0.08
G1 X100.0 E6  F1000.0 ; fat +40mm intro line @ 0.15
G1 E-0.8 F3000; retract to avoid stringing
G1 X99.5 E0 F1000.0 ; -0.5mm wipe action to avoid string
G1 X110.0 E0 F1000.0 ; +10mm intro line @ 0.00
G1 E0.6 F1500; de-retract
G92 E0.0 ; reset extrusion distance
; Final print adjustments `

So if I dont get you solution to work I could also just manually edit 20 gcodes and get them in the correct order
